File: templates/problems/17_math_geometry/02_spiral_matrix/spiral_matrix_test.go
package spiral_matrix

import (
	"reflect"
	"testing"
)

func TestSpiralOrder(t *testing.T) {
	tests := []struct {
		name     string
		matrix   [][]int
		expected []int
	}{
		{"3x3", [][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}, []int{1, 2, 3, 6, 9, 8, 7, 4, 5}},
		{"3x4", [][]int{{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10, 11, 12}}, []int{1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7}},
		{"1x1", [][]int{{1}}, []int{1}},
		{"1 row", [][]int{{1, 2, 3}}, []int{1, 2, 3}},
		{"1 column", [][]int{{1}, {2}, {3}}, []int{1, 2, 3}},
		{"2x2", [][]int{{1, 2}, {3, 4}}, []int{1, 2, 4, 3}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := spiralOrder(tt.matrix)
			if !reflect.DeepEqual(got, tt.expected) {
				t.Errorf("spiralOrder(%v) = %v, want %v", tt.matrix, got, tt.expected)
			}
		})
	}
}

File: templates/problems/17_math_geometry/02_spiral_matrix/typescript/spiralMatrix.test.ts
import { describe, it, expect } from 'vitest';
import { spiralOrder } from './spiralMatrix';

describe('SpiralMatrix', () => {
    it('3x3', () => {
        expect(spiralOrder([[1, 2, 3], [4, 5, 6], [7, 8, 9]])).toEqual([1, 2, 3, 6, 9, 8, 7, 4, 5]);
    });

    it('3x4', () => {
        expect(spiralOrder([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])).toEqual([1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7]);
    });

    it('1x1', () => {
        expect(spiralOrder([[1]])).toEqual([1]);
    });

    it('1 row', () => {
        expect(spiralOrder([[1, 2, 3]])).toEqual([1, 2, 3]);
    });

    it('1 column', () => {
        expect(spiralOrder([[1], [2], [3]])).toEqual([1, 2, 3]);
    });

    it('2x2', () => {
        expect(spiralOrder([[1, 2], [3, 4]])).toEqual([1, 2, 4, 3]);
    });
});
